R/proj.R

```diff
@@ -0,0 +1,249 @@
+#'   or character vector indicating specific draws may also be provided.
+#'
+#' @references
+#' McCartan, C. (2024). Projective Averages for Summarizing Redistricting
+#' Ensembles. *arXiv preprint*. Available at \url{https://arxiv.org/abs/}.
+#'
+#' @examples
+#' data(iowa)
+#' map <- redist_map(iowa, existing_plan = cd_2010, pop_tol = 0.01)
+#' plans <- redist_smc(map, 50, silent = TRUE)
+#' plans$dem <- group_frac(map, dem_08, tot_08, plans)
+#'
+#' proj_distr(plans, dem) # a 99-by-50 matrix
+#' plot(map, proj_avg(plans, dem))
+#' plot(map, proj_contr(plans, dem))
+#' plot(map, proj_contr(plans, dem, comp="cd_2010"))
+#'
+#' @concept analyze
+#' @name proj
+NULL
+
+
+#' @returns `proj_distr`: A matrix with a row for each precinct (row in the map
+#'   object) and a column for every draw described by `draws`.
+#'
+#' @rdname proj
+#' @export
+proj_distr <- function(plans, x, draws=NA) {
+    plans_m <- get_plans_matrix(plans)
+
+    n_ref <- 0
+    # copied from get_n_ref()
+    if (!is.null(colnames(plans_m))) {
+        n_ref <- sum(nchar(colnames(plans_m)) > 0)
+    }
+
+    if (is.null(draws)) {
+        draw_idx <- seq_len(ncol(plans_m))
+    } else if (length(draws) == 1 && is.na(draws)) {
+        if (n_ref > 0) {
+            draw_idx <- seq_len(ncol(plans_m))[-seq_len(n_ref)]
+        } else {
+            draw_idx <- seq_len(ncol(plans_m))
+        }
+    } else if (is.logical(draws)) {
+        draw_idx <- which(draws)
+    } else {
+        draw_idx <- match(as.character(draws), levels(plans$draw))
+    }
+
+    plans <- vctrs::vec_slice(plans, vctrs::vec_order(cbind(plans$draw, plans$district)))
+    n_distr <- attr(plans, "ndists")
+    x <- rlang::eval_tidy(rlang::enquo(x), plans)
+
+    proj_distr_m(plans_m, x, draw_idx, n_distr)
+}
+
+#' @returns `proj_avg`: A numeric vector of length matching the number of
+#'   precincts.
+#'
+#' @rdname proj
+#' @export
+proj_avg <- function(plans, x, draws=NA) {
+    rowMeans(proj_distr(plans, {{ x }}, draws))
+}
+
+#' @param comp The plan to compare to the rest of the ensemble (which is
+#'   controlled by `draws`). Defaults to the first reference plan, if any exists
+#' @param norm If `TRUE`, normalize the contrast by the standard deviation of
+#'   the projective distribution, precinct-wise. This will make the projective
+#'   contrast in terms of z-scores.
+#' @param pfdr If `TRUE`, calculate q-values for each precinct that can be
+#'   used to control the positive false discovery rate (pFDR) at a given level
+#'   by thresholding the q-values at that level. Q-values are stored as the
+#'   `"q"` attribute on the returned vector. Requires the `matrixStats` package
+#'   be installed.
+#'
+#' @returns `proj_contr`: A numeric vector of length matching the number of
+#'   precincts, optionally with a `"q"` attribute containing q-values.
+#'
+#' @rdname proj
+#' @export
+proj_contr <- function(plans, x, comp=NA, draws=NA, norm=FALSE, pfdr=FALSE) {
```

Comment:
Some naming stuff: 
`comp` = competitiveness some days, `comp` = compactness other days, and `comp` = comparison too. Can we call this `comparison` or something like that? 

and `norm` -> `normalize`? 

And `pfdr` -> `p_fdr` to match the other fn?

Reply:
`pFDR` is one abbreviation so shouldn't get an underscore